src/aspire/classification/align2d.py

```diff
@@ -225,35 +376,620 @@ def align(self, classes, reflections, basis_coefficients):
+        if self.alignment_src.L != source.L:
+            raise RuntimeError("Currently `alignment_src.L` must equal `source.L`")
+        if self.alignment_src.dtype != source.dtype:
+            raise RuntimeError(
+                "Currently `alignment_src.dtype` must equal `source.dtype`"
+            )
+
+        # Sanity check. This API should be rethought once all basis and
+        # alignment methods have been incorporated.
+        assert alignment_basis is None  # We use sources directly for alignment
+        assert (
+            composite_basis is not None
+        )  # However, we require a basis for rotating etc.
+
+        super().__init__(
+            alignment_basis, source, composite_basis, batch_size=batch_size, dtype=dtype
+        )
+
+    def _phase_cross_correlation(self, img0, img1):
+        """
+        # Adapted from skimage.registration.phase_cross_correlation
+
+        :param img0: Fixed image.
+        :param img1: Translated image.
+        :returns: (cross-correlation magnitudes (2D array), shifts)
+        """
+
+        # Cache img0 transform, this saves n_classes*(n_nbor-1) transforms
+        # Note we use the `id` because ndarray are unhashable
+        src_f = self.__cache.setdefault(id(img0), np.fft.fft2(img0))
+
+        target_f = np.fft.fft2(img1)
+
+        # Whole-pixel shifts - Compute cross-correlation by an IFFT
+        shape = src_f.shape
+        image_product = src_f * target_f.conj()
+        cross_correlation = np.fft.ifft2(image_product)
+
+        # Locate maximum
+        maxima = np.unravel_index(
+            np.argmax(np.abs(cross_correlation)), cross_correlation.shape
+        )
+        midpoints = np.array([np.fix(axis_size / 2) for axis_size in shape])
+
+        shifts = np.array(maxima, dtype=np.float64)
+        shifts[shifts > midpoints] -= np.array(shape)[shifts > midpoints]
+
+        return np.abs(cross_correlation), shifts
+
+    def _align(self, classes, reflections, basis_coefficients):
+        """
+        Performs the actual rotational alignment estimation,
+        returning parameters needed for averaging.
+        """
+
+        # Admit simple case of single case alignment
+        classes = np.atleast_2d(classes)
+        reflections = np.atleast_2d(reflections)
+
+        n_classes = classes.shape[0]
+
+        # Instantiate matrices for results
+        rotations = np.zeros(classes.shape, dtype=self.dtype)
+        correlations = np.zeros(classes.shape, dtype=self.dtype)
+        shifts = np.zeros((*classes.shape, 2), dtype=int)
+
+        for k in trange(n_classes):
+            # # Get the array of images for this class, using the `alignment_src`.
+            images = self._cls_images(classes[k], src=self.alignment_src)
+
+            self._reddychatterji(
+                k, images, classes, reflections, rotations, correlations, shifts
+            )
+
+        return classes, reflections, rotations, shifts, correlations
+
+    def _reddychatterji(
```

Comment:
These output arguments threw me for a bit. Can we just have them returned instead? Then we should be able to get rid of `k` as an argument as well.

Reply:
`_reddychatterji`  is intentionally operating in place, so it has no output args.

Reply:
If I'm understanding correctly here, `rotations`, `correlations`, and `shifts` are passed in and modified by the function, while `classes` and `reflections` are not. For all these arguments, only the `k`th element is accessed, so wouldn't it be simpler to call it like
```
rotations[k], correlations[k], shifts[k] = _reddychatterji(images, classes[k], reflections[k])
```
Is there a speed advantage to in-place operation here?

Reply:
Thanks!

Reply:
This is done.